File: Solid_Principles/Open_Closed_Principle/02_OCP.cpp
#include <iostream>

// TITLE: Open/Closed Principle (OCP) - InvoiceDao Example

class Invoice {
public:
    double amount;
    Invoice(double amt) : amount(amt) {}
};

// ========================================
// THE PROBLEM (VIOLATION):
// To add MongoDB, we must MODIFY this class
// ========================================

class InvoiceDaoBad {
public:
    void saveToDB(Invoice inv) {
        std::cout << "Saving to MySQL DB" << std::endl;
    }
    
    void saveToFile(Invoice inv) {
        std::cout << "Saving to File" << std::endl;
    }
    
    // Adding new method = MODIFYING class (violates OCP!)
    void saveToMongoDB(Invoice inv) {
        std::cout << "Saving to MongoDB" << std::endl;
    }
};

// ========================================
// THE SOLUTION:
// Use interface - extend via NEW classes
// ========================================

class InvoiceDao {
public:
    virtual void save(Invoice inv) = 0;
    virtual ~InvoiceDao() {}
};

class DatabaseInvoiceDao : public InvoiceDao {
public:
    void save(Invoice inv) override {
        std::cout << "Saving to MySQL Database" << std::endl;
    }
};

class FileInvoiceDao : public InvoiceDao {
public:
    void save(Invoice inv) override {
        std::cout << "Saving to File System" << std::endl;
    }
};

// Adding MongoDB doesn't modify existing code!
class MongoDBInvoiceDao : public InvoiceDao {
public:
    void save(Invoice inv) override {
        std::cout << "Saving to MongoDB" << std::endl;
    }
};

int main() {
    Invoice inv(100.0);
    
    std::cout << "=== PROBLEM (OCP Violation) ===" << std::endl;
    InvoiceDaoBad badDao;
    badDao.saveToDB(inv);
    badDao.saveToFile(inv);
    badDao.saveToMongoDB(inv);
    
    std::cout << "\n=== SOLUTION (OCP Compliant) ===" << std::endl;
    InvoiceDao* dbDao = new DatabaseInvoiceDao();
    dbDao->save(inv);
    
    InvoiceDao* fileDao = new FileInvoiceDao();
    fileDao->save(inv);
    
    InvoiceDao* mongoDao = new MongoDBInvoiceDao();
    mongoDao->save(inv);
    
    delete dbDao;
    delete fileDao;
    delete mongoDao;
    
    return 0;
}

File: Solid_Principles/Open_Closed_Principle/02_OCP.py
from abc import ABC, abstractmethod

# TITLE: Open/Closed Principle (OCP) - InvoiceDao Example

class Invoice:
    def __init__(self, amount):
        self.amount = amount

# ========================================
# THE PROBLEM (VIOLATION):
# To add MongoDB, we must MODIFY this class
# ========================================

class InvoiceDaoBad:
    def save_to_db(self, invoice):
        print("Saving to MySQL DB")
    
    def save_to_file(self, invoice):
        print("Saving to File")
    
    # Adding new method = MODIFYING class (violates OCP!)
    def save_to_mongodb(self, invoice):
        print("Saving to MongoDB")

# ========================================
# THE SOLUTION:
# Use abstraction - extend via NEW classes
# ========================================

class InvoiceDao(ABC):
    @abstractmethod
    def save(self, invoice):
        pass

class DatabaseInvoiceDao(InvoiceDao):
    def save(self, invoice):
        print("Saving to MySQL Database")

class FileInvoiceDao(InvoiceDao):
    def save(self, invoice):
        print("Saving to File System")

class MongoDBInvoiceDao(InvoiceDao):
    def save(self, invoice):
        print("Saving to MongoDB")

if __name__ == "__main__":
    inv = Invoice(100.0)
    
    print("=== PROBLEM (OCP Violation) ===")
    bad_dao = InvoiceDaoBad()
    bad_dao.save_to_db(inv)
    bad_dao.save_to_file(inv)
    bad_dao.save_to_mongodb(inv)
    
    print("\n=== SOLUTION (OCP Compliant) ===")
    db_dao = DatabaseInvoiceDao()
    db_dao.save(inv)
    
    file_dao = FileInvoiceDao()
    file_dao.save(inv)
    
    mongo_dao = MongoDBInvoiceDao()
    mongo_dao.save(inv)
